### PR TITLE
[ops, perf] refactor: compute MoE scatter-index in O(N) instead of two sorts

### PR DESCRIPTION
**Pull request by zhangxin81**
### What does this PR do?

The Triton and Quack fused-MoE backends built the dispatch scatter-index with a
back-to-back `argsort(stable=True).argsort()`. The second `argsort` only inverts
a permutation of `[0..N)` — an inherently O(N) operation that was implemented as
a full O(N log N) sort, i.e. one extra sort-kernel launch per MoE forward and per
MoE backward, per layer. The source already carried a `TODO(wenyawei): opt it`
next to this line.

This PR introduces a shared helper, `compute_expert_scatter_index`, that keeps the
single stable sort and builds the inverse permutation directly with one
`arange` + scatter, then rewires all three call sites to use it. Behavior is
preserved bit-exact.

### Checklist Before Starting

- Search for relative PRs/issues and link here: addresses the in-code
  `TODO(wenyawei): opt it` at the MoE scatter-index construction.
- PR title follows `[{modules}] {type}: {description}` format — `[ops, perf] refactor: ...`.

### Test

`tests/ops/test_expert_scatter_index.py` (new, 10 cases, all passing on CPU):

```
$ pytest tests/ops/test_expert_scatter_index.py -v
...
============================= 10 passed in 32.79s ==============================
```

Coverage:
- Bit-exact parity with the previous `argsort(stable=True).argsort().int()`
  expression across 6 `(num_tokens, num_experts, topk)` shapes.
- `scatter_index` is a permutation of `[0, N)`.
- Stability: same-expert tokens keep their original `(token, top-k slot)` order
  (the property group-GEMM relies on).
- Dtype/device preserved: `sorted_order` int64, `scatter_index` int32, same device
  as input.
- `sorted_order` and `scatter_index` are mutual inverses.

`make quality` passes (ruff check + format).

#### GPU benchmark (H100)

Measured with a standalone microbenchmark (not part of this PR) comparing the old
`argsort(stable=True).argsort()` expression against `compute_expert_scatter_index`
across realistic MoE shapes.

- Device: `NVIDIA H100 80GB HBM3` (SM 9.0); PyTorch 2.7.1 / CUDA 12.8; dtype `int64`
- Timing: `torch.cuda.Event`, 20 warmup + 200 timed iters per measurement,
  3 interleaved rounds, reported as median of medians
- Numerical parity (`torch.equal(old, new)`) verified per shape before timing;
  all 10 shapes passed

|     T | topk | experts | old (ms) | new (ms) | speedup |
| ----: | ---: | ------: | -------: | -------: | ------: |
|  4096 |    8 |     128 |   0.2010 |   0.1117 |   1.80x |
|  8192 |    8 |     128 |   0.2103 |   0.1160 |   1.81x |
| 16384 |    8 |     128 |   0.2238 |   0.1227 |   1.82x |
| 32768 |    8 |     128 |   0.2487 |   0.1375 |   1.81x |
| 65536 |    8 |     128 |   0.2864 |   0.1618 |   1.77x |
|  4096 |    4 |      64 |   0.1921 |   0.1061 |   1.81x |
| 16384 |    4 |      64 |   0.2100 |   0.1159 |   1.81x |
| 65536 |    4 |      64 |   0.2487 |   0.1375 |   1.81x |
|  4096 |    2 |       8 |   0.1957 |   0.1069 |   1.83x |
| 16384 |    2 |       8 |   0.2000 |   0.1109 |   1.80x |

Consistent 1.77x–1.83x speedup across every tested shape. The win comes from
eliminating one full `argsort` kernel launch (the second sort); what remains is
one sort + one O(N) scatter. Absolute per-call time is small (~0.1–0.3 ms), but
MoE forward+backward hits this dispatch on every layer, so the removed launch
compounds across a step.

### API and Usage Example

No public API change. Internal helper only:

```python
# veomni/ops/kernels/moe/_scatter.py
def compute_expert_scatter_index(expert_index: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
    """Return (sorted_order, scatter_index) for MoE dispatch."""
```

### Design & Code Changes

Root cause — `sorted_order = flat.argsort(stable=True)` returns a permutation of
`[0..N)`; its inverse is computable in O(N) via a single scatter, but the code
re-used `argsort` for it, incurring a full O(N log N) sort of that permutation.

Fix — add `veomni/ops/kernels/moe/_scatter.py`:

```python
flat = expert_index.flatten()
sorted_order = flat.argsort(stable=True)
inv = torch.empty_like(sorted_order)
inv[sorted_order] = torch.arange(sorted_order.numel(), dtype=sorted_order.dtype, device=sorted_order.device)
scatter_index = inv.to(torch.int32).view(expert_index.shape)
return sorted_order, scatter_index
```

Properties preserved so downstream kernels don't notice the change:
- **Stability** — `stable=True` on the remaining sort keeps same-expert tokens in
  original flat-index order.
- **Dtype** — returns int32 (the Triton `moe_scatter`/`moe_gather` kernels take
  int32 indices).
- **`sorted_order` pass-through** — the Quack path needs it for
  `A_idx = sorted_order // topk`, so the helper returns it (no recomputation).
- **Device-agnostic** — derives device from the input tensor; no `torch.cuda.*`.

Call sites rewired (3):
- `veomni/ops/kernels/moe/group_gemm.py` — `TritonFusedMoeExpertFunction.forward`
- `veomni/ops/kernels/moe/group_gemm.py` — `MergedFc1TritonFusedMoeExpertFunction.forward`
- `veomni/ops/kernels/moe/quack_gemm.py` — `_build_moe_indices`

Files changed:

| Path | Kind | Delta |
| --- | --- | --- |
| `veomni/ops/kernels/moe/_scatter.py` | new | +79 |
| `veomni/ops/kernels/moe/group_gemm.py` | modified | +6 -3 |
| `veomni/ops/kernels/moe/quack_gemm.py` | modified | +5 -3 |
| `tests/ops/test_expert_scatter_index.py` | new | +112 |

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks (`make quality` passes)
- Added/updated documentation — N/A (no public API, config, or workflow change)
- If `tasks/` training scripts were moved or renamed — N/A
- Added tests to CI workflow — `tests/ops/test_expert_scatter_index.py` (CPU-only,
  runs in the standard `pytest tests/` collection)

### CODE REVIEW

**Review by Luosuu — APPROVED** [review]

**Review by Coach257 — APPROVED** [review]

**Review by FoolPlayer — APPROVED** [review]
LGTM. Reviewed the MoE scatter-index refactor: the new helper preserves the stable argsort ordering and inverse-permutation semantics while avoiding the second O(N log N) sort; call sites and CPU parity tests look consistent.
